[Ang Wei Jun] iP

aweijun opened the pull request:
# Duke

> "Use GitHub Flavored Markdown" - CS2103T [[Source]](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html).

This is a simple task list that helps your to noted down all your different task.

* Text Based
* Easy to Use
* ~~Super~~**Extremely** convenient to use

## How to start?
1. Download it from [here](https://github.com/aweijun/ip/releases/tag/A-Jar).
2. Use java to launch the application. 
3. Add your task using the given commands. 
4. Access it anytime.

#### Duke Commands
```
todo taskname
deadline taskname /by --
event taskname /at --
bye
```
#### Duke Features
- [x] Adding and deleting different tasks `todo <task>` `deadline <task> /by <time>` `event <task> /at <time>`
- [x] Finding task based on their name `find`
- [ ] Storing task into a local storage file
#### Unsolved errors 👎 
- [ ] Add more OOP for the storage class 
- [ ] Ensure that errors are thrown by the appropriate classes

Commit:
Add Gradle support

Commit:
Level-1

Commit:
Level 2 Commit

Commit:
Level-3 Commit

aweijun committed:
Level 4 Commit

Commit:
Level 5 commit

aweijun committed:
Level-6 commit

aweijun committed:
Level-6 update

aweijun committed:
Level-6 Update

aweijun committed:
Fixed Bug involving exceptions thrown

aweijun committed:
Added the storing functionality, encountered error:
Exception in thread "main" java.lang.UnsupportedClassVersionError: Duke has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up
 to 52.0

aweijun committed:
Solved the unsafe errors

aweijun committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/ToDo.java

aweijun committed:
Add DateTime Functionality to Deadline and Event Task.

aweijun committed:
Merge branch 'branch-Level-8'

aweijun committed:
Add missing modifiers

aweijun committed:
Modify files to Ui, Parser, Storage and TaskList

aweijun committed:
Separate to Ui, Storage, Parser, TaskList,
Add files into relevant packages

aweijun committed:
Add Command Class - Separate Command into different sub-classes
Edit Parser Class - Add getCommand function

aweijun committed:
Fix minor errors MarkCommand
Edit Storage loadFromDisk function

aweijun committed:
Add Junits for DeadLine and ToDo.

No testing was implemented.

Add test for Deadline and ToDo task using Junit. Tested the display
and toStore methods.

Testing using Junit would allow the automated testing of said
functions

aweijun committed:
Add new DukeExceptions, Account for minor bugs related to Mark,
Unmark and delete, task adding commands

Ran into out of bounds and datetime parser e related to mark,
unmark and delete and task commands.

Add new exceptions to catch said errors

This is to account for any possible user errors

aweijun committed:
Remove chatbot class and moved logic to Duke. Standardize the
welcome message into a Response class.

The logic was not consistent previously.

This would allow for better organization of the code

aweijun committed:
Add basic documentation.

The lack of java documentation helps users to understand the
purpose of different classes and functions.

Add documentation to various classes and methods.

aweijun committed:
Fix multiple coding violations

Removal of unnecessary this statements.
Make array names plural
Change switch-case format

aweijun committed:
Add FindCommand and FindResponse.

Edit Parser to take in find command, as per the requirements of
Level-9

aweijun committed:
Merge branch 'Level-9'

aweijun committed:
Merge branch 'A-JavaDoc'

aweijun committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeadlineCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/data/Storage.java
#	src/main/java/duke/data/TaskList.java
#	src/main/java/duke/parser/Parser.java

aweijun committed:
Fixed coding violations

Fixed coding violations across all the files

aweijun committed:
Merge branch 'add-gradle-support' of https://github.com/aweijun/ip into A-Gradle

aweijun committed:
Change the package statements for the files.

The package statement was incorrect after the addition of Gradle.

Fixed the issue by removing main.java in front of package statements.

aweijun committed:
Merge branch 'A-Gradle'

aweijun committed:
Add checkstyle to Duke and fix coding violations

Fixed coding violations in Duke.

aweijun committed:
Merge branch 'A-CheckStyle'

thikhinab commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+     * Constructor for Duke Chatbot
+     */
+    public Duke() {
+        this.taskList = new TaskList();
+        this.store = new Storage(taskList);
+        this.commandHandler = new Parser();
+        this.cmdLine = new Ui();
```
Would it be less redundant if `this` keyword was not used?
```suggestion
        taskList = new TaskList();
        store = new Storage(taskList);
        commandHandler = new Parser();
        cmdLine = new Ui();
```

thikhinab commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+     */
+    public void run() {
+        store.initialiseStorage();
+        taskList = store.loadFromDisk();
```
Great choice in naming methods!

thikhinab commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+        while (isRunning) {
+            try {
+                String stringCmd = cmdLine.getNextLine();
+                Command cmd = commandHandler.getCommand(stringCmd);
```
Would it be clearer if `cmd` was explicitly mentioned as `command`?

thikhinab commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,110 @@
+package duke.data;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.Scanner;
+
+import duke.dukeexceptions.DukeException;
+import duke.dukeexceptions.ForeignException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.ToDo;
```
Great work in explicitly mentioning all imported classes!

thikhinab commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    private static boolean storageToMark(String mark) {
+        if (mark.equals("1")) {
+            return true;
+        } else {
+            return false;
+        }
```
```suggestion
        return mark.equals("1");
```

thikhinab commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,110 @@
+
+    private static Task storageToTask(String taskString) {
+        String[] stringCmdUnits = taskString.split(" \\| ");
+        switch (stringCmdUnits[0]) {
+        case "T":
+            return new ToDo(stringCmdUnits[2], storageToMark(stringCmdUnits[1]));
+        case "D":
+            return new Deadline(stringCmdUnits[2], storageToMark(stringCmdUnits[1]), stringCmdUnits[3]);
+        case "E":
+            return new Event(stringCmdUnits[2], storageToMark(stringCmdUnits[1]), stringCmdUnits[3]);
+        default:
+            return null;
```
Good use in switch statements and good formatting!

thikhinab commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,110 @@
+            }
+            return taskList;
+        } catch (Exception e) {
+            e.getMessage();
```
What is `e.message()` used for?

thikhinab commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,69 @@
+ * Parser that deals with user inputs.
+ */
+public class Parser {
+    /***
```
Would the formatting look better if there is space between line 23 and 24?

thikhinab commented on src/main/java/duke/responses/FindResponse.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class FindResponse implements Response {
+
+    private TaskList iterate;
```
Would it be easy to understand if `iterate` was renamed as `tasks`?

thikhinab commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+        this.name = name;
+    }
+    /***
+     * Void methods that sets the completeness of the task based on the input.
```
Should the JavaDoc comment of the method start with a third-person present tense? e.g. `Sets the completeness of the task based on the input.`

thikhinab reviewed (commented):
Overall, great work in code styling! Only a few minor suggestions.

aweijun committed:
Implemented the GUI for Level-10.

Change the return value of the exception callback function into a
String. Add the GUI components into Duke.

aweijun committed:
Merge branch 'Level-10'

BrendonLau commented on src/main/java/duke/responses/AddTaskResponse.java:
```diff
@@ -0,0 +1,36 @@
+
+public class AddTaskResponse implements Response {
+    private Task currTask;
+    private TaskList tasklist;
```
For naming convention is it better to use taskList?

BrendonLau reviewed (commented):
Other than what the other reviewer mentioned, LGTM.

aweijun committed:
Add JUnit for Parser. Fixed minor UI related.

Add extra JUnit for Parser class checking the parse function.
Fix minor UI errors observed.
Added a case for list command

aweijun committed:
Fix bug involving storage.

Storage was not saved and load upon that of the Bot.

aweijun committed:
Fix bug regarding storage.

The storage was not updating upon delete command.
Updated the storage after each command.

arcornior commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    public void initialiseStorage() {
+        File data = new File("data");
+        try {
+            if (data.exists()) {
+            } else {
+                data.mkdirs();
+            }
+            if (persistStore.exists()) {
+            } else {
+                persistStore.createNewFile();
+            }
+        } catch (Exception e) {
+            e.getMessage();
+        }
```
Great job maximising the readability of your code!

arcornior commented on src/main/java/duke/data/TaskList.java:
```diff
@@ -0,0 +1,77 @@
+     * Marks the task at index to be done
+     * @param index index in the task
+     */
+    public void markTask(int index) {
+        tasks.get(index).setMark(true);
+    }
+
+    /**
+     * Marks the task at index to be undone
+     * @param index index in the task
+     */
+    public void unmarkTask(int index) {
+        tasks.get(index).setMark(false);
+    }
```
Would it be better to merge these functions to minimise code duplication?

arcornior commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,69 @@
+        String[] stringCmdUnits = stringCmd.split(" ");
+        String stringFirstCmd = stringCmdUnits[0];
+        switch (stringFirstCmd) {
+        case "bye":
+            return new ByeCommand();
+        case "delete":
+            return new DeleteCommand(stringCmd);
+        case "list":
+            return new ListCommand();
+        case "mark":
+            return new MarkCommand(stringCmd);
+        case "unmark":
+            return new UnmarkCommand(stringCmd);
+        case "todo":
+            return new TodoCommand(stringCmd);
+        case "deadline":
+            return new DeadlineCommand(stringCmd);
+        case "event":
+            return new EventCommand(stringCmd);
+        case "find":
+            return new FindCommand(stringCmd);
+        default:
+            throw new ForeignException("");
```
Good job extracting the logic for each command into separate classes!

arcornior commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+
+    private static boolean storageToMark(String mark) {
```
Would changing the name to convertMarkToBoolean be better?

arcornior commented on src/main/java/duke/data/Storage.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    /**
+     * function that loads from permanent storage to the TaskList.
```
Should function be capitalised?
```suggestion
     * Function that loads from permanent storage to the TaskList.
```

arcornior reviewed (commented):
LGTM! Great job maximising code readability!

awj99 committed:
Fix GUI Bye command.

Bye command was not working.
Update duke application in order to close after bye command.

awj99 committed:
Fix GUI list command.

Fixed GUI list command not stating the whole list.

awj99 committed:
Add assertions.

Assertions are used to ensure that the length of the TaskList increase
in length by 1 after each time the todo, deadline, even commands are
executed.

Assertions are used to define assumptions about the program state so
that the runtime can verify them. An assertion failure indicates a
possible bug in the code.

aweijun committed:
Merge branch 'master' into branch-A-Assertions

aweijun committed:
Merge pull request #1 from aweijun/branch-A-Assertions

Add assertions

awj99 committed:
Change minor code quality. Remove testing printf statements.

Used the KISS to cut down on redundant code in Storage.java.
Cleaned up printf statements to improve the readability of the code.

awj99 committed:
Merge branch 'master' of https://github.com/aweijun/ip

aweijun committed:
Merge pull request #2 from aweijun/branch-A-CodeQuality

Change minor code quality

awj99 committed:
Implement duplicated task feature

Duplicated task may cause confusion to users as it is difficult to
differentiate tasks of the same name.

Ensures that users do not add duplicated task with the same name
to the task list. In the event that when a duplicated task is added, a
warning is given instead. Informing the user that there is already a task
with the same name in the task list.

This feature aims to improve user experience by reducing user error.

awj99 committed:
Merge branch 'branch-C-DetectDuplicates'

aweijun committed:
Create github-actions-demo.yml

aweijun committed:
Merge pull request #4 from aweijun/aweijun-patch-1

Create github-actions-demo.yml

aweijun committed:
Set theme jekyll-theme-hacker

awj99 committed:
Merge branch 'master' of https://github.com/aweijun/ip

awj99 committed:
Added User guide

Update Readme to contain all the necessary information for user guide.
This includes listing out all the features and the usage of each of
said features.

User guides contains important information about the product. Such as the
set up process, available functions or usage of the products etc.

awj99 committed:
Solve bug regarding storage.

Fix bug where the date and time was not properly stored into the storage.

awj99 committed:
Create the jar using Gradle

Perform minor edits to build.gradle to facilitate the making of the jar
using gradle.

awj99 committed:
Fix minor Bug in Deadline

Fix bug where the information within deadline was not stored properly.

awj99 committed:
Final submission

Create final jar file for the submission.
